### ISSUE TITLE
datalad push has concurrency/locking issues

### ISSUE THREAD

**Issue by mih** [datalad/datalad#4656]
Scenario: hundreds of jobs change different clones of the same dataset. They save the generated content into different branches and `push`. Random pushes fail at like this:

```
publish(ok): . (dataset) [refs/heads/condor-60585.2->origin:refs/heads/condor-60585.2 [new branch]]
copy(ok): testfunny (file) [to origin...]
publish(error): . (dataset) [refs/heads/git-annex->origin:refs/heads/git-annex [rejected] (fetch first)]
Potential hints to solve encountered errors: 
1: Updates were rejected because the remote contains work that you do not have locally. This is usually caused by another repository pushing to the same ref. You may want to first integrate the remote changes (e.g., 'git pull ...') before pushing again. See the 'Note about fast-forwards' in 'git push --help' for details.
```

i.e. `git-annex` branch is gets outdated -- I think with the `push` call.

**Comment by mih** [comment]
I am thinking about wrapping the `datalad push` call in the analysis script with `flock` pointing to a global lock.

**Comment by bpoldrack** [comment]
I am thinking about branching `git-annex` and merging it at the end. Not well thought-through in terms of workflow (who does the merge and when), but want to put a trace of that thought here.

**Comment by mih** [comment]
I think it is happening between the sync and push within the pish call.

**Comment by mih** [comment]
I did not know that a `clone` can fail, when a `push` isn't complete yet -- but it does. So it really needs a global repo lock covering all operations.

**Comment by bpoldrack** [comment]
> I did not know that a clone can fail, when a push isn't complete yet -- but it does. So it really needs a global repo lock covering all operations.

Ah, that didn't occur to me. Sounds like a good point. Too bad, though. I kinda liked the idea to just branch out git-annex the same way we'd do for the result branches. :-/

**Comment by yarikoptic** [comment]
What git-annex version was that?  I wonder if it relates to the recent change in git-annex behavior causing a commit on remote end, which was addressed in the publish tests in https://github.com/datalad/datalad/pull/4623

**Comment by mih** [comment]
7.20190819+git2-g908476a9b-1~ndall+1

**Comment by yarikoptic** [comment]
have you seen this one recently @mih ?

**Comment by mih** [comment]
No, I have not. But may be due to a change in the standard workflow setup that clones from A but always pushes to B, and leave an ultimate merge and consolidation of be to a subsequent finalization stage.

I am OK to close this issue.
